Add some type hinting to the classes inside models.

Serhiy1 opened the pull request:
Completes type hinting in the test suite structure #4570

The files that got type hints added, where mainly files that I have interacted with as an end user. 

The PR is fully Compatible with Compatible with python 3.6 onwards.

pekkaklarck commented on src/robot/model/body.py:
```diff
@@ -122,7 +130,7 @@ def create(self):
         return self._create(self.keyword_class, 'create_keyword', args, kwargs)
 
-    def _create(self, cls, name, args, kwargs):
+    def _create(self, cls:Union[Type[C], None], name: str, args, kwargs) -> C:
```
- `cls` can only be `None` is there's a bug in our code so it shouldn't be included in type hints.
- This method creates body items so I believe this should be `_create(self, cls: Type[BodyItem], name: str, args: list, kwargs: dict) -> BodyItem`
- This is mostly internal code so I'm not sure are type hints needed in general. If the above works, it may be simple enough to be useful, but I believe more complex type hints with `TypeVar` aren't worth the effort.

pekkaklarck commented on src/robot/model/body.py:
```diff
@@ -222,15 +230,15 @@ class Body(BaseBody):
     __slots__ = ['branch_class']
 
-    def __init__(self, branch_class, parent=None, items=None):
+    def __init__(self, branch_class: Type[C], parent=None, items=None):
```
`branch_class` cannot be anything (i.e. generic). It actually only be `IfBranch` or `TryBranch`, but just using `BodyItem` could be fine as well.

pekkaklarck commented on src/robot/model/control.py:
```diff
@@ -14,10 +14,10 @@
 
+from typing import Any, Dict
 
 @Body.register
```
- Standard module imports should be before internal imports.
- Our style is to have an empty row between absolute imports from top level module (`from robot.xxx import ...`) and relative imports from the same package (`from .xxx import ...`).
- Two empty lines needed before class declaration (and a class decorator is part of the declaration).

pekkaklarck commented on src/robot/model/control.py:
```diff
@@ -117,7 +117,7 @@ def _include_in_repr(self, name, value):
 
     def to_dict(self):
-        data = {'type': self.type}
+        data: Dict[str, Any] = {'type': self.type}
```
This doesn't look too useful, especially inside a method. It would be better to add a type hint to the method, but I believe just `dict` is good enough unless we'd use a `TypedDict` to properly document the structure. Using a `TypedDict` with all `to_dict` methods would be a huge task and not in the scope of this PR. It would probably be better to use something like pydantic for documenting the `to_dict/json` and `from_dict/json` methods anyway.

pekkaklarck commented on src/robot/model/control.py:
```diff
@@ -313,6 +313,7 @@ class Return(BodyItem):
     type = BodyItem.RETURN
     repr_args = ('values',)
     __slots__ = ['values']
+    body: Body
```
`Return`, `Continue` and `Break` don't have a `body` except on the result side.

pekkaklarck commented on src/robot/model/fixture.py:
```diff
@@ -15,10 +15,16 @@
+def create_fixture(fixture: Union["F", None], parent: object, fixture_type: str) -> "F":
     # TestCase and TestSuite have 'fixture_class', Keyword doesn't.
-    fixture_class = getattr(parent, 'fixture_class', parent.__class__)
+    fixture_class: Type["F"] = getattr(parent, 'fixture_class', parent.__class__)
```
This looks _really_ complicated for a method that's only used internally. If this was to be typed, `@overload` would probably help, but I believe it's not worth the effort.

pekkaklarck commented on src/robot/model/keyword.py:
```diff
@@ -15,9 +15,14 @@
 
 import warnings
 
+
```
Remove.

pekkaklarck commented on src/robot/model/keyword.py:
```diff
@@ -15,9 +15,14 @@
 
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from robot.model.visitor import SuiteVisitor
```
Standard library import in wrong place.

pekkaklarck commented on src/robot/model/keyword.py:
```diff
@@ -29,12 +34,15 @@ class Keyword(BodyItem):
 
-    def __init__(self, name='', args=(), assign=(), type=BodyItem.KEYWORD, parent=None):
+
+    def __init__(self, name:str='', args: tuple=(), assign: tuple=(), type=BodyItem.KEYWORD, parent=None):
```
- `name:str=''` -> `name: str = ''`
- Should add type hints to all parameters.
- `args` and `assign` probably could also be lists, but I believe `tuple` is fine as a type hint. Can be changed to something more generic like `Sequence` later if needed.

pekkaklarck commented on src/robot/model/keyword.py:
```diff
@@ -29,12 +34,15 @@ class Keyword(BodyItem):
+        self.teardown: "Keyword"
 
     @property
     def name(self):
```
Return type hint missing.

pekkaklarck commented on src/robot/model/keyword.py:
```diff
@@ -65,7 +73,7 @@ def to_dict(self):
 
 
-class Keywords(ItemList):
+class Keywords(ItemList[Keyword]):
```
Although this class is called `Keywords`, it actually contains `BodyItems`. Not a huge deal either way, it will be removed in RF 7.

pekkaklarck commented on src/robot/model/testcase.py:
```diff
@@ -34,8 +39,11 @@ class TestCase(ModelObject):
-    def __init__(self, name='', doc='', tags=None, timeout=None, lineno=None,
-                 parent=None):
+    def __init__(self, name: str = '', doc: str = '',
+                 tags: Union[List[str], str, None] = None,
```
`tags` can also be `Tags` instance. It could also be a tuple or any other iterable but I'm not sure is that relevant.

pekkaklarck commented on src/robot/model/testcase.py:
```diff
@@ -86,7 +94,7 @@ def setup(self):
 
     @setup.setter
-    def setup(self, setup):
+    def setup(self, setup: Keyword):
```
`setup` can also be `None`.

pekkaklarck commented on src/robot/model/testcase.py:
```diff
@@ -114,7 +122,7 @@ def teardown(self):
 
     @teardown.setter
-    def teardown(self, teardown):
+    def teardown(self, teardown: Keyword):
```
Can also be `None`.

pekkaklarck commented on src/robot/model/testcase.py:
```diff
@@ -162,15 +170,15 @@ def longname(self):
     def source(self):
```
Return type hint missing here and possibly elsewhere.

pekkaklarck commented on src/robot/model/testcase.py:
```diff
@@ -180,21 +188,21 @@ def to_dict(self):
+        for visitor in test.parent._visitors:   #type: ignore
             test.visit(visitor)
-        return test
+        return test
```
Files should end with a newline.

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -40,8 +44,11 @@ class TestSuite(ModelObject):
-    def __init__(self, name: str = '', doc: str = '', metadata: dict = None,
-                 source: Path = None, rpa: bool = False, parent: 'TestSuite' = None):
+    def __init__(self, name: str = '', doc: str = '',
+                 metadata: Union[dict, None] = None,
```
`metadata` can also be `Metadata`.

Serhiy1 replied:
Fixed

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -50,12 +57,12 @@ def __init__(self, name: str = '', doc: str = '', metadata: dict = None,
-        self._setup = None
-        self._teardown = None
+        self._setup: Union[Keyword, None] = None
+        self._teardown: Union[Keyword,  None] = None
```
Not part of the public API so type hints probably aren't needed.

Serhiy1 replied:
Removed

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -50,12 +57,12 @@ def __init__(self, name: str = '', doc: str = '', metadata: dict = None,
 
     @staticmethod
-    def name_from_source(source: Path):
+    def name_from_source(source: Union[Path, str, None]):
```
Return type missing.

Serhiy1 replied:
Fixed

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -83,37 +90,37 @@ def name(self):
 
     @property
-    def setup(self):
+    def setup(self) -> Union[Keyword, None]:
```
`setup` and `teardown` are never `None` when accessed.

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -142,7 +149,7 @@ def setup(self):
 
     @setup.setter
-    def setup(self, setup):
+    def setup(self, setup: Keyword) -> Union[Keyword, None]:
```
- Passed in `setup` can be `None`.
- This method doesn't return anything.

pekkaklarck commented on src/robot/model/visitor.py:
```diff
@@ -159,10 +159,10 @@ def visit_test(self, test: TestCase):
             test.body.visit(self)
             if test.has_teardown:
-                test.teardown.visit(self)
+                test.teardown.visit(self) # type: ignore
```
I don't like `# type: ignore` comments in general, but in these cases they shouldn't even be needed.

Serhiy1 replied:
I will remove these. Unfortunately, most IDEs don't understand that there is a valid check beforehand.  has_teardown and has_setup will need to implement [typeguards](https://peps.python.org/pep-0647) but they are not available until the very latest versions of python

pekkaklarck replied:
`test.teardown` and others ought to have type hint `Keyword` so there shouldn't be errors calling `test.teardown.visit(self)`. At least earlier the type hint was `Keyword | None` which would have issues.

The `has_teardown` check is there to avoid accessing `test.teardown` property that creates the underlying `Keyword` object. It wouldn't be a problem otherwise, but creating lot of such objects unnecessarily consumes memory.

pekkaklarck reviewed (changes requested):
Great work but lot of things to enhance still. General problems:
- Some style violations especially related to imports.
- In some case types added to attributes that don't even exist.
- Return types and also parameter types missing in many places in public APIs.
- Some of the internal type hints look so complicated that they may not be useful.

Fixing everything in this PR and then re-reviewing is a pretty big task. I believe it would be best to start from a single module like `robot.model.testsuite` and type it thoroughly. Once we get that done, we set standard that can then be followed when typing other modules.

Serhiy1 commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,35 @@
+
+	// Uncomment to connect as root instead. More info: https://aka.ms/dev-containers-non-root.
+	// "remoteUser": "root"
+}
```
This was accidentally added removing

Serhiy1 commented:
> Fixing everything in this PR and then re-reviewing is a pretty big task. I believe it would be best to start from a single module like robot.model.testsuite and type it thoroughly. Once we get that done, we set standard that can then be followed when typing other modules.

@pekkaklarck I think I have gone through all the requested changes inside `robot.model.testsuite`

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -27,6 +27,10 @@
+from typing import Any, TYPE_CHECKING, Dict, List, Union
+if TYPE_CHECKING:
+    from robot.model.visitor import SuiteVisitor
+
```
- Standard module imports should be before internal imports.
- We discussed on #devel about using `'x | y'` `'list[x]'` style instead of  `Union[x, y]` and `list[x]`. That would avoid lot of `from typing` imports, although in this case we anyway needed to import `TYPE_CHECKING`.

Serhiy1 replied:
Should I switch to stringifying all the type hints in the file?

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -40,8 +44,11 @@ class TestSuite(ModelObject):
+                 metadata: Union[Metadata, None] = None,
+                 source: Union[Path, None] = None,
+                 rpa: bool = False,
+                 parent: Union["TestSuite", None] = None):
```
- `metadata` can also be `dict` and  `source` can also be `str`.
- Using stringly types like `'Metadata | dict | None'` would probably be better than using `Union[Metadata, dict, None]`.
- Please use single quotes instead of double quotes.
- We need to decide how to handle long signatures like this one. Splitting all parameters to own lines would be a simple solution, but it should be done consistently (i.e. also `name` and `doc` on own lines) and such formatting takes lot of vertical space. I'd try having multiple params per line, but we can make a final decision a bit later. Notice that we use 88 as the maximum line length.

pekkaklarck commented on src/robot/model/testsuite.py:
```diff
@@ -302,7 +315,7 @@ def __str__(self):
 
     def to_dict(self):
-        data = {'name': self.name}
+        data: Dict[str, Any] = {'name': self.name}
```
I don't think this helps. The method should get `-> dict`, though.

Serhiy1 replied:
I added the type hint to suppress some reported type errors from VScode, with the first assignment, it assumes that keys and values for the dict are always strings raising errors all over the place. 

Will remove though.

pekkaklarck commented:
> @pekkaklarck I think I have gone through all the requested changes inside `robot.model.testsuite`

I added few more comments. I mentioned two things to decide in them:
1. Should we use stringified types like `'x | y'` and `'list[int]'` or use `Union`, `List`, etc? We discussed this on #devel and most people seemed to prefer the former. I also like that style more, especially with unions, and it avoids lot of `from typing` imports.
2. How to format long signatures. Having each parameter on its own line would be an easy solution, but such style takes lot of vertical space. I believe we should try fitting multiple params per line.

How do you feel about the above? If you agree with me, please update `testsuite` to use such style. We can then see how it feels and decide should we use that style also with other files.

The next question is how to merge only `testsuite` now. I think I cannot, at least here on web, merge only one file, so a new PR would probably be easiest.

Serhiy1 commented:
Changes to `TestSuite.py`  
- Re-ordered imports
- Switched to stringified type hints 
- Changed init signature to maximise horizontal character count before newlines

As for, just emerging TestSuite.py, I will roll back and locally stash my changes and use them in a follow-up PR
